### PR TITLE
[WIP] Issue/164 - Added support to trusted servers to wakeup a device

### PR DESCRIPTION
**Pull request by frsela**

### CODE REVIEW

**Comment by frsela** [comment]
Related issue #164

**Comment by frsela** [comment]
Note: After land PR #164 (issue #163) I will add client certificate check for trusted servers (certificates signed by us)

File: src/ns_as/as_server.js [hunk 1]
```diff
@@ -240,6 +240,50 @@ server.prototype = {
       });
       break;
 
+    case 'wakeup':
+      // TODO: Check if trusted server
```
Comment (willyaranda) [inline src/ns_as/as_server.js:244]: Do you mean we have some kind of trusted servers on the ddbb?

This should be some kind of "partnership"?

File: src/ns_as/as_server.js [hunk 2]
```diff
@@ -240,6 +240,50 @@ server.prototype = {
...
+      if (request.method != 'POST') {
+        log.debug('NS_AS::onHTTPMessage --> No valid method (only POST for wakeup)');
+        return response.res(errorcodesAS.BAD_URL_NOT_VALID_METHOD);
+      }
+
+      log.debug('NS_AS::onHTTPMessage --> Wakeup . . .');
+      request.on('data', function(data) {
+        try {
+          var json = JSON.parse(data);
+          log.info("1");
```
Comment (willyaranda) [inline src/ns_as/as_server.js]: remove :)

File: src/ns_as/as_server.js (hunk 1 above)
Comment (frsela) [inline src/ns_as/as_server.js:244]: No, more easy.

After PR 166 (cert auth support) I will check here if the received client certificate is one SIGNED by our CA so external CAs will be used for notifications and our CA for privilaged API.

We can sign a client certificate for all partners we want to allow them to wakeup handsets in our networks.

WDYT?

File: src/ns_as/as_server.js (hunk 2 above)
Comment (frsela) [inline src/ns_as/as_server.js]: Ups ... :)

File: src/ns_as/as_server.js [hunk 3]
```diff
@@ -240,6 +240,50 @@ server.prototype = {
...
+              "mobilenetwork": {
+                "mcc": json.wakeup.mobilenetwork.mcc,
+                "mnc": json.wakeup.mobilenetwork.mnc
+              }
+            },
+            "messageType": "wakeUP_TrustedServer"
+          });
+        } catch (s) {
+          return response.res(errorcodesAS.JSON_NOTVALID_ERROR);
+        }
```
Comment (willyaranda) [inline src/ns_as/as_server.js:256]: I would rewrote this to be something like:

```
try {
  var json = JSON.parse(data);
} catch (e) {
  return response.res(errorcodesAS.JSON_NOTVALID_ERROR);
}
//check the json, return early if is not valid

//send to the msgBroker.
```

I think the try-catch block is too big and not too beautiful to read.

Comment (frsela) [inline src/ns_as/as_server.js:256]: agree. changing

File: src/ns_as/as_server.js (hunk 1 above)
Comment (willyaranda) [inline src/ns_as/as_server.js:244]: Check with @antonioma ;) I don't know if this is something I can talk about :)

File: src/ns_as/as_server.js [hunk 4]
```diff
@@ -240,6 +240,50 @@ server.prototype = {
...
+        return response.res(errorcodesAS.BAD_URL_NOT_VALID_METHOD);
+      }
+
+      log.debug('NS_AS::onHTTPMessage --> Wakeup . . .');
+      request.on('data', function(data) {
+        try {
+          var json = JSON.parse(data);
+        } catch (s) {
+          return response.res(errorcodesAS.JSON_NOTVALID_ERROR);
+        }
```
Comment (willyaranda) [inline src/ns_as/as_server.js:256]: Much more cleaner, cool!

File: src/ns_as/as_server.js (hunk 1 above)
Comment (AntonioMA) [inline src/ns_as/as_server.js:244]: Hmm... We can find a synergy with BV here. They're using certificates also for some (all?) client apps and we could use their infrastructure for that. But... let's finish what we have before getting into new wars ;)

**Comment by willyaranda** [comment]
I think we need to stop this PR until more decision is taken. I like it: it can be used by other services as you point out!!

File: src/ns_as/as_server.js (hunk 4 above)
Comment (frsela) [inline src/ns_as/as_server.js:256]: agree ;)

**Comment by frsela** [comment]
Yes, it's not finished yet . . . is more a WIP ;) Thanks willy.

File: src/ns_as/as_server.js (hunk 1 above)
Comment (frsela) [inline src/ns_as/as_server.js:244]: Move discussion to mumble/skype, it's faster ;)

Comment (frsela) [inline src/ns_as/as_server.js:244]: After the skype chat, we agree on the use of trusted CA and client certs for trusted servers.

This PR shall be merged after PR 163 and rebased and added CA verifications ;)

**Comment by frsela** [comment]
Invalid, this will be moved to an independent project.

**Pull request closed without merging**